### PR TITLE
Runtime: move cleaning of state from expression to step

### LINKED ISSUE #827: Logger: Removed configuration from final state
I hate this line:
```
Removed configuration from final state
```

It's a reasonably important line. It runs at the end of every step.

But when there's an error, it tends to appear before the error. Or it's the last  log line we receive.

And it looks  like it's the cause of a fail, rather than the effect of a step being completed.

I'm not really sure what to do about this. Logging to debug and adding a debug filter would help.

### PR DESCRIPTION
**Pull request by doc-han**
## Short Description
This PR makes sure that cleanup of state happens after the report of where a job failed or passed. 
<table><thead>
  <tr>
    <th>Job Failed</th>
    <th>Job Successful</th>
  </tr></thead>
<tbody>
  <tr>
    <td></td>
    <td></td>
  </tr>
</tbody>
</table>

1. Moves the cleaning of state after a job expression has been executed to the step level
2. Moves tests for state cleaning to step
3. Updates step state cloning to use safe-stringify

Fixes #827

### Before this PR:
Cleanup of state to remove additional keys like `configuration` and those provided through `statePropsToRemove` happens at the job expression execution level. Hence, cleanup logs happen before job execution status is logged. 

### After this PR:
Cleanup of state has been moved to the step execution level which isn't that far from job expression execution. Hence, cleanup logs can happen after job execution status is logged.

### Note ⛔️⛔️⛔️
1. Tests that check cleaning of state in `expression.test.ts` have been moved to `step.test.ts`
2. The test for circular objects in `expression.test.ts` was passing fine but didn't catch for when same was passed to a step. This is because these objects where used by a clone function before the expression was executed which will make it fail.
https://github.com/OpenFn/kit/blob/6b6ff2d2d8cd8078856aad9506ae3bcf70ac88cf/packages/runtime/test/execute/expression.test.ts#L86
3. The test below was a pseudo test. The object used for the test was the same used for assertion and since it was passed by reference it was passing when it was supposed to fail.
https://github.com/OpenFn/kit/blob/6b6ff2d2d8cd8078856aad9506ae3bcf70ac88cf/packages/runtime/test/execute/expression.test.ts#L146-L157

## QA Notes

List any considerations/cases/advice for testing/QA here.

## AI Usage

Please disclose how you've used AI in this work (it's cool, we just want to know!):

- [ ] Code generation (copilot but not intellisense)
- [ ] Learning or fact checking
- [ ] Strategy / design
- [ ] Optimisation / refactoring
- [ ] Translation / spellchecking / doc gen
- [ ] Other
- [x] I have not used AI

You can read more details in our [Responsible AI Policy](https://www.openfn.org/ai#pull-request-templates)

I understand this might be too much for an initial PR. happy to discuss this PR

### CODE REVIEW

**Comment by doc-han** [comment]
An additional commit to move the result of a failed job below the actual error to deal with the confusion. 

Screenshot 2024-11-20 at 3 05 18 PM

**Comment by doc-han** [comment]
### Flaky test
https://github.com/OpenFn/kit/blob/6b6ff2d2d8cd8078856aad9506ae3bcf70ac88cf/packages/engine-multi/test/integration.test.ts#L140

**Comment by josephjclark** [comment]
Thanks @doc-han! Taking a look now.

That flaky test is covered here: https://github.com/OpenFn/kit/issues/761

**Comment by josephjclark** [comment]
@doc-han Actually I think I prefer your first screenshot, where we first say "woah there's been an error!" and then print the error

File: packages/runtime/src/execute/step.ts [hunk 1]
```diff
@@ -138,15 +169,21 @@ const executeStep = async (
...
+        const { error, state: errState } = e as ExecutionErrorWrapper;
+        let state = errState;
 
         const duration = logger.timer(timerId);
-        logger.error(`Failed step ${jobName} after ${duration}`);
+
         report(state, jobId, error);
 
+        logger.error(
+          `Step (${jobName}) aborted with status: Error (after ${duration})`
```
[the remark this answers is not in the document]
Comment (doc-han) [reply packages/runtime/src/execute/step.ts:180]: alright

**Comment by doc-han** [comment]
> @doc-han Actually I think I prefer your first screenshot, where we first say "woah there's been an error!" and then print the error
> 
> image
> 
> I find that way more intuitive - can we revert to it?

Yeah. I'll remove the second commit then.

**Comment by doc-han** [comment]
> I appreciate the thoroughness of the tests and I'm happy with the changes. A lot of those runtime tests need refactoring because it's testing way too broad a slice of code. I just didn't have time to refactor it nicely back in the day.
> 
> One more request: can you please run `pnpm changeset` and generate a release note? This will be publicly visible but not by the kind of people who like to fork runtimes. So please keep it brief and high level. Check the runtime changelog for inspiration.
> 
> Thank you so much for this!

Alright. I'm glad you're okay with the change. 🚀

**Comment by josephjclark** [comment]
@doc-han Don't forget to run `pnpm changeset` please! 

If you're not sure how to handle it let me know and I'll merge and do it for you.

**Comment by doc-han** [comment]
@josephjclark just added a commit for the changelog.

1. I'm just confused whether I'm supposed to manually update the version to `1.5.2`
2. a new file to .changeset, is that expected?

**Comment by doc-han** [comment]
> @josephjclark just added a commit for the changelog.
> 
> 1. I'm just confused whether I'm supposed to manually update the version to `1.5.2`
> 2. a new file to .changeset, is that expected?

I can remove the commit if it doesn't fit so this can be merged.

**Comment by josephjclark** [comment]
Hey @doc-han - sorry for the late response. Super busy day!

You've done it perfectly: that changelog file is exactly what I need. `changeset` will bump the versions for us later on.

I'll get this merged and released - thank you!!

**Comment by josephjclark** [comment]
@doc-han Checkout https://github.com/OpenFn/kit/pull/830 and note all the `package.json`s and changelogs that got updated. That all came from the single changeset file :sunglasses:

**Comment by doc-han** [comment]
> @doc-han Checkout #830 and note all the `package.json`s and changelogs that got updated. That all came from the single changeset file 😎

Beautiful!

**Pull request merged**